src/components/tl/sharp/tl_sharp_context.c

```diff
@@ -0,0 +1,369 @@
+{
+    ucc_tl_sharp_context_t *ctx          = (ucc_tl_sharp_context_t*)context;
+    ucc_tl_sharp_rcache_region_t *region = ucs_derived_of(rregion,
+                                                ucc_tl_sharp_rcache_region_t);
+    int ret;
+
+    ret = sharp_coll_dereg_mr(ctx->sharp_context, region->reg.mr);
+    if (ret < 0) {
+        tl_error(ctx->super.super.lib, "dereg faiedi(%d). mr:%p",
+                 ret, region->reg.mr);
+    } else {
+        tl_debug(ctx->super.super.lib, "rregion:%p dere mr:%p",
+                 rregion, region->reg.mr);
+    }
+}
+
+static void
+ucc_tl_sharp_rcache_dump_region_cb(void *context, ucs_rcache_t *rcache,
+                                   ucs_rcache_region_t *rregion, char *buf,
+                                   size_t max)
+{
+    ucc_tl_sharp_rcache_region_t *region = ucs_derived_of(rregion,
+                                           ucc_tl_sharp_rcache_region_t);
+
+    snprintf(buf, max, "bar ptr:%p", region->reg.mr);
+}
+
+static ucc_rcache_ops_t ucc_tl_sharp_rcache_ops = {
+    .mem_reg     = ucc_tl_sharp_rcache_mem_reg_cb,
+    .mem_dereg   = ucc_tl_sharp_rcache_mem_dereg_cb,
+    .dump_region = ucc_tl_sharp_rcache_dump_region_cb
+};
+
+UCC_CLASS_INIT_FUNC(ucc_tl_sharp_context_t,
+                    const ucc_base_context_params_t *params,
+                    const ucc_base_config_t *config)
+{
+    ucc_tl_sharp_context_config_t *tl_sharp_config =
+        ucc_derived_of(config, ucc_tl_sharp_context_config_t);
+    struct sharp_coll_init_spec init_spec = {0};
+    ucc_status_t status;
+    ucc_rcache_params_t rcache_params;
+
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_context_t, tl_sharp_config->super.tl_lib,
+                              params->context);
+    memcpy(&self->cfg, tl_sharp_config, sizeof(*tl_sharp_config));
+
+    self->rcache           = NULL;
+    self->oob_ctx.ctx      = self;
+    self->oob_ctx.ctx_oob  = &UCC_TL_CTX_OOB(self);
+    self->oob_ctx.team_oob = NULL;
+
+    init_spec.progress_func                  = NULL;
+    init_spec.world_rank                     = UCC_TL_CTX_OOB(self).oob_ep;
+    init_spec.world_local_rank               = 0;
+    init_spec.world_size                     = UCC_TL_CTX_OOB(self).n_oob_eps;
+    init_spec.enable_thread_support          = 1;
+    init_spec.group_channel_idx              = 0;
+    init_spec.oob_colls.barrier              = ucc_tl_sharp_oob_barrier;
+    init_spec.oob_colls.bcast                = ucc_tl_sharp_oob_bcast;
+    init_spec.oob_colls.gather               = ucc_tl_sharp_oob_gather;
+    init_spec.oob_ctx                        = &self->oob_ctx;
+    init_spec.config                         = sharp_coll_default_config;
+    init_spec.config.user_progress_num_polls = 1000000;
+    init_spec.config.ib_dev_list             = self->cfg.dev_list;
+    init_spec.job_id                         = ((getpid() ^ pthread_self())
+                                                ^ rand());
```

Comment:
we should not use default "rand" (with default user seed) - this may break some user apps. lets use rand_r with ucc internal (or even tl sharp local) seed